cpp/Documentation.md

```diff
@@ -0,0 +1,474 @@
+
+.. automodule:: hello
+    :members:
+    :undoc-members:
+    :show-inheritance:
+```
+
+This is enough to generate documentation of the `hello` module.
+To generate documentation of symbols imported by `hello` module, consider using
+the `:imported-members:` option of the `automodule` command.
+
+
+#### Integrate documentation of C++ code
+
+```diff
+diff --git a/doc/source/conf.py b/doc/source/conf.py
+index 4b51de3..f1109d9 100644
+--- a/doc/source/conf.py
++++ b/doc/source/conf.py
+@@ -39,6 +39,8 @@ release = '0.0.1'
+ # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
+ # ones.
+ extensions = [
++    'breathe',
++    'exhale',
+     'sphinx.ext.autodoc',
+     'sphinx.ext.doctest',
+     'sphinx.ext.coverage',
+@@ -47,6 +49,27 @@ extensions = [
+     'sphinx.ext.githubpages',
+ ]
+
++# Setup the breathe extension
++breathe_projects = {
++    "Basalt C++ Library": "./doxyoutput/xml"
++}
++breathe_default_project = "Basalt C++ Library"
++
++# Setup the exhale extension
++exhale_args = {
++    # These arguments are required
++    "containmentFolder":     "./cpp_api",
++    "rootFileName":          "library_root.rst",
++    "rootFileTitle":         "C++ API",
++    "doxygenStripFromPath":  "..",
++    # Suggested optional arguments
++    "createTreeView":        True,
++    # TIP: if using the sphinx-bootstrap-theme, you need
++    # "treeViewIsBootstrap": True,
++    "exhaleExecutesDoxygen": True,
++    # additional Doxygen config
++    "exhaleDoxygenStdin":    "INPUT = ../include"
++}
++
+ # Add any paths that contain templates here, relative to this directory.
+ templates_path = ['_templates']
+
+diff --git a/doc/source/index.rst b/doc/source/index.rst
+index d218dc5..8c1cdd0 100644
+--- a/doc/source/index.rst
++++ b/doc/source/index.rst
+@@ -11,6 +11,7 @@ Welcome to Basalt's documentation!
+    :caption: Contents:
+
+    modules.rst
++   cpp_api/library_root
+
+```
+
+### Toward a decent `setup.py`
+
+Sphinx provides a `build_sphinx` setuptools target to generate documentation
+with the command: `python setup.py build_sphinx`.
+
+#### Setup dependencies
+
+Add all Python packages required to build the documentation to the `setup_requires`
+[*setup* keyword](https://setuptools.readthedocs.io/en/latest/setuptools.html#new-and-changed-setup-keywords)
+
+If `doctest` is activated, then the generation of the documentation executes the Python package. Thus it is
+recommended to also add the content of `install_requires` *setup* keyword to the `setup_requires` keyword.
+
+For instance:
+
+```python
+install_requirements = [
+    "cached-property>=1.5.1",
+    "docopt>=0.6.2",
+    "h5py>=2.7.1",
+    "humanize>=0.5.1",
+    "numpy>=1.13",
+    "progress>=1.4",
+]
+doc_requirements = [
+    "exhale",
+    "m2r",
+    "sphinx-rtd-theme",
+    "sphinx<2"
+]
+setup(
+    # [...]
+    install_requires=install_requirements,
+    setup_requires=install_requirements + doc_requirements
+)
+```
+
+#### Test command on steroids
```

Comment:
more of a meta comment. While this is certainly a very useful feature and an interesting approach to build new features into your project (by starting with the doc and an example and then adding the implemention) I would not promote this to **the way** to setup a python project (sorry, if that was not the intention). I would therefore mark it as optional.

Reply:
I didn't want to force every C++ project to be a Python package.
I clarified this in cbce428